helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java

```diff
@@ -122,115 +122,166 @@ public InstanceInfo getInstanceInfo(String clusterId, String instanceName,
-          clusterId, instanceName, e);
+      LOG.error("Failed to perform custom client side instance level health checks for {}/{}", clusterId, instanceName,
+          e);
       throw e;
     }
   }
 
-  @VisibleForTesting
-  protected StoppableCheck performPartitionLevelChecks(String clusterId, String instanceName,
-      String baseUrl, Map<String, String> customPayLoads) throws IOException {
-    LOG.info("Perform partition level health checks for {}/{}", clusterId, instanceName);
-    // pull the health status from ZK
+  private Map<String, StoppableCheck> performPartitionsCheck(List<String> instances, RESTConfig restConfig,
+      Map<String, String> customPayLoads) {
     PartitionHealth clusterPartitionsHealth = generatePartitionHealthMapFromZK();
-    Map<String, List<String>> expiredPartitionsOnInstances =
-        clusterPartitionsHealth.getExpiredRecords();
     // update the health status for those expired partitions on instances
-    try {
-      for (Map.Entry<String, List<String>> entry : expiredPartitionsOnInstances.entrySet()) {
-        Map<String, Boolean> partitionHealthStatus =
-            _customRestClient.getPartitionStoppableCheck(baseUrl, entry.getValue(), customPayLoads);
-        partitionHealthStatus.entrySet().forEach(kv -> clusterPartitionsHealth
-            .updatePartitionHealth(instanceName, kv.getKey(), kv.getValue()));
-      }
-    } catch (IOException e) {
-      LOG.error("Failed to perform client side partition level health checks for {}/{}", clusterId,
-          instanceName, e);
-      throw e;
+    Map<String, List<String>> expiredPartitionsByInstance = clusterPartitionsHealth.getExpiredRecords();
+    Map<String, Future<Map<String, Boolean>>> updatedPartitionsHealthStatusByInstance = new HashMap<>();
+    for (Map.Entry<String, List<String>> entry : expiredPartitionsByInstance.entrySet()) {
+      String instance = entry.getKey();
+      List<String> expiredPartitions = entry.getValue();
+      Callable<Map<String, Boolean>> refreshTask =
+          () -> _customRestClient.getPartitionStoppableCheck(getBaseUrl(instance, restConfig), expiredPartitions,
+              customPayLoads);
+      updatedPartitionsHealthStatusByInstance.put(instance, POOL.submit(refreshTask));
     }
+    updatedPartitionsHealthStatusByInstance.entrySet().forEach(instanceToPartitionStatus -> {
+      String instance = instanceToPartitionStatus.getKey();
+      try {
+        Map<String, Boolean> updatedPartitionsHealth = instanceToPartitionStatus.getValue().get();
+        updatedPartitionsHealth.entrySet()
+            .forEach(partitionToHealthStatus -> clusterPartitionsHealth.updatePartitionHealth(instance,
+                partitionToHealthStatus.getKey(), partitionToHealthStatus.getValue()));
+      } catch (InterruptedException | ExecutionException e) {
+        LOG.error("Failed to get partition status on instance {}", instance, e);
+      }
+    });
```

Comment:
I would suggest to wrap these ZK + REST call for partition level update to a separate function.

Reply:
If you don't mind, I would move the `getPartitionHealthMap` method into the accessorWrapper class to eliminate the business logics in the class. It will enlarge the PR (so I initially put a TODO only)

Reply:
That will be good.